File: src/Spreadsheet/DataMap.fs
module ARCtrl.Spreadsheet.Datamap

open ARCtrl
open ArcTable
open FsSpreadsheet

/// Reads an datamap from a spreadsheet
let fromFsWorkbook (doc : FsWorkbook) = 
    try
        let worksheets = doc.GetWorksheets()
        let sheetIsEmpty (sheet : FsWorksheet) = sheet.CellCollection.Count = 0
        let dataMapTable = 
            worksheets
            |> Seq.tryPick DatamapTable.tryFromFsWorksheet
        match dataMapTable with
        | Some table -> table
        | None -> 
            if worksheets |> Seq.forall sheetIsEmpty then
                Datamap.init()
            else
                failwith "No DatamapTable was found in any of the sheets of the workbook"
    with
    | err -> failwithf "Could not parse datamap: \n%s" err.Message
            
let toFsWorkbook (dataMap : Datamap) =
    let doc = new FsWorkbook()

    DatamapTable.toFsWorksheet dataMap
    |> doc.AddWorksheet
    doc

File: js/Spreadsheet/DataMap.js
import { printf, toFail } from "../fable_modules/fable-library-js.4.22.0/String.js";
import { tryPick } from "../fable_modules/fable-library-js.4.22.0/Seq.js";
import { toFsWorksheet, tryFromFsWorksheet } from "./DataMapTable/DataMapTable.js";
import { FsWorkbook } from "../fable_modules/FsSpreadsheet.6.3.0-alpha.4/FsWorkbook.fs.js";

/**
 * Reads an datamap from a spreadsheet
 */
export function fromFsWorkbook(doc) {
    try {
        const dataMapTable = tryPick(tryFromFsWorksheet, doc.GetWorksheets());
        if (dataMapTable == null) {
            throw new Error("No DataMapTable found in workbook");
        }
        else {
            return dataMapTable;
        }
    }
    catch (err) {
        const arg = err.message;
        return toFail(printf("Could not parse datamap: \n%s"))(arg);
    }
}

export function toFsWorkbook(dataMap) {
    const doc = new FsWorkbook();
    const sheet = toFsWorksheet(dataMap);
    doc.AddWorksheet(sheet);
    return doc;
}

